### PR TITLE
[InstCombine] Remove shl if we only demand known signbits of shift source

### PR DESCRIPTION
**Pull request by ParkHanbum**
this patch resolve TODO written in commit
https://github.com/llvm/llvm-project/commit/5909c678831f3a5c1669f6906f777d4ec4532fa1

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp [hunk 1]
```diff
@@ -640,25 +640,31 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
                                                     DemandedMask, Known))
             return R;
 
-      // TODO: If we only want bits that already match the signbit then we don't
+      uint64_t ShiftAmt = SA->getLimitedValue(BitWidth - 1);
+      // If we only want bits that already match the signbit then we don't
       // need to shift.
+      if (DemandedMask.countr_zero() >= ShiftAmt) {
+        unsigned NumLowDemandedBits = BitWidth - DemandedMask.countr_zero();
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:647]: Isn't this highbits?

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp [hunk 2]
```diff
@@ -640,25 +640,31 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
...
+        unsigned SignBits =
+            ComputeNumSignBits(I->getOperand(0), Depth + 1, CxtI);
+        if (SignBits > ShiftAmt && SignBits - ShiftAmt >= NumLowDemandedBits)
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:650]: This can be `>=`:
https://alive2.llvm.org/ce/z/C3VNoR

Comment (ParkHanbum) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:650]: @goldsteinn maybe your link is wrong. 
anyway. considering that if signbits equal shiftamt, 
https://alive2.llvm.org/ce/z/82A769
It doesn't seem to work.

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp (hunk 1 above)
Comment (ParkHanbum) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:647]: my bad.. I'll fix it!

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp [hunk 3]
```diff
@@ -640,25 +640,31 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
                                                     DemandedMask, Known))
             return R;
 
-      // TODO: If we only want bits that already match the signbit then we don't
+      uint64_t ShiftAmt = SA->getLimitedValue(BitWidth - 1);
+      // If we only want bits that already match the signbit then we don't
       // need to shift.
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:645]: Move this comment inside the if.

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp [hunk 4]
```diff
@@ -640,25 +640,31 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
...
-        Constant *NewC = ConstantFoldBinaryOpOperands(Instruction::Shl, C,
-                                                      LeftShiftAmtC, DL);
-        if (ConstantFoldBinaryOpOperands(Instruction::LShr, NewC, LeftShiftAmtC,
-                                         DL) == C) {
-          Instruction *Lshr = BinaryOperator::CreateLShr(NewC, X);
-          return InsertNewInstWith(Lshr, I->getIterator());
+        // If we can pre-shift a right-shifted constant to the left without
+        // losing any high bits amd we don't demand the low bits, then eliminate
+        // the left-shift:
+        // (C >> X) << LeftShiftAmtC --> (C << RightShiftAmtC) >> X
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:656]:
```suggestion
        // losing any high bits and we don't demand the low bits, then eliminate
        // the left-shift:
        // (C >> X) << LeftShiftAmtC --> (C << LeftShiftAmtC) >> X
```
Pre-existing, but if you're touching the code anyway...

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp (hunk 2 above)
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp]: I think that @goldsteinn is in principle correct, but if `SignBits == ShiftAmt` then `SignBits - ShiftAmt` is 0, so the condition would become `0 >= NumHighDemandedBits`, which is the same as `NumHighDemandedBits == 0`, which is a degenerate case that gets handled earlier anyway. So this just doesn't matter either way.

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp (hunk 3 above)
Comment (ParkHanbum) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:644]: sorry my miss!

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp (hunk 2 above)
Comment (ParkHanbum) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp]: I'm update alive link for showing verification of case `SignBits == ShiftAmt` and `SignBits > ShiftAmt && SignBits - ShiftAmt >= NumHighDemandedBits`.  if I misunderstand of it, please let me know.

**Comment by dtcxzyw** [comment]
@ParkHanbum Can you rebase this patch on main?

**Comment by ParkHanbum** [comment]
@dtcxzyw it is done!

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp [hunk 5]
```diff
@@ -655,25 +655,31 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
...
       }
 
-      // TODO: If we only want bits that already match the signbit then we don't
+      // We only want bits that already match the signbit then we don't
       // need to shift.
+      uint64_t ShiftAmt = SA->getLimitedValue(BitWidth - 1);
+      if (DemandedMask.countr_zero() >= ShiftAmt) {
+        unsigned NumHiDemandedBits = BitWidth - DemandedMask.countr_zero();
+        unsigned SignBits =
+            ComputeNumSignBits(I->getOperand(0), Depth + 1, CxtI);
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:664]: Can we just check `nsw` here? We will set `nsw` if `computeNumSignBits >= ShCnt` + other conditions.

Comment (ParkHanbum) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:664]: It seems that knowing that `computeNumSignBits` is greater than `ShCnt` is not enough. We need to shift `computeNumSignBits` by `ShCnt` and then check whether `DemandedBits` is still within the range of `SignBits`. 
If there is a way to check the range of `SignBits` and `DemandedBits` by checking `nsw`, please let me know.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:664]: Ah you are right, although you should be able to check `nsw` first before doing the `computeNumSignBits` as w.o `nsw` think we always know `ShCnt > computeNumSignBits`.

Comment (ParkHanbum) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:664]: @goldsteinn I understand!

**Comment by goldsteinn** [comment]
LGTM

**Pull request merged**